File: 23/25/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
	"sync"
	"sync/atomic"
)

// Graph represents an undirected graph
type Graph struct {
	Edges map[string]map[string]bool
}

// NewGraph initializes a new graph
func NewGraph() *Graph {
	return &Graph{Edges: make(map[string]map[string]bool)}
}

// AddEdge adds an edge to the graph
func (g *Graph) AddEdge(node1, node2 string) {
	if g.Edges[node1] == nil {
		g.Edges[node1] = make(map[string]bool)
	}
	if g.Edges[node2] == nil {
		g.Edges[node2] = make(map[string]bool)
	}
	g.Edges[node1][node2] = true
	g.Edges[node2][node1] = true
}

// RemoveEdge removes an edge from the graph
func (g *Graph) RemoveEdge(node1, node2 string) {
	delete(g.Edges[node1], node2)
	delete(g.Edges[node2], node1)
}

// Copy creates a deep copy of the graph
func (g *Graph) Copy() *Graph {
	copy := NewGraph()
	for node, neighbors := range g.Edges {
		for neighbor := range neighbors {
			copy.AddEdge(node, neighbor)
		}
	}
	return copy
}

// GetConnectedComponents returns the sizes of connected components
func (g *Graph) GetConnectedComponents() []int {
	visited := make(map[string]bool)
	var componentSizes []int

	var dfs func(string) int
	dfs = func(node string) int {
		stack := []string{node}
		size := 0
		for len(stack) > 0 {
			curr := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if visited[curr] {
				continue
			}
			visited[curr] = true
			size++
			for neighbor := range g.Edges[curr] {
				if !visited[neighbor] {
					stack = append(stack, neighbor)
				}
			}
		}
		return size
	}

	for node := range g.Edges {
		if !visited[node] {
			componentSizes = append(componentSizes, dfs(node))
		}
	}
	return componentSizes
}

// CalculateGroupProduct calculates the product of the sizes of two largest components
func (g *Graph) CalculateGroupProduct(edgesToRemove [][2]string) int {
	copy := g.Copy()
	for _, edge := range edgesToRemove {
		copy.RemoveEdge(edge[0], edge[1])
	}
	componentSizes := copy.GetConnectedComponents()
	if len(componentSizes) < 2 {
		return 0
	}
	return componentSizes[0] * componentSizes[1]
}

// ReadGraphFromFile loads a graph from a file
func ReadGraphFromFile(filePath string) *Graph {
	graph := NewGraph()
	file, err := os.Open(filePath)
	if err != nil {
		log.Fatalf("Error opening file: %v", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ":")
		node := strings.TrimSpace(parts[0])
		neighbors := strings.Fields(parts[1])
		for _, neighbor := range neighbors {
			graph.AddEdge(node, neighbor)
		}
	}
	return graph
}

// GenerateEdgeCombinations generates all 3-edge combinations
func GenerateEdgeCombinations(edges [][2]string) [][3][2]string {
	var combinations [][3][2]string
	n := len(edges)
	for i := 0; i < n-2; i++ {
		for j := i + 1; j < n-1; j++ {
			for k := j + 1; k < n; k++ {
				combinations = append(combinations, [3][2]string{edges[i], edges[j], edges[k]})
			}
		}
	}
	return combinations
}

func worker(graph *Graph, tasks <-chan [3][2]string, results chan<- [2]interface{}, progress *int32) {
	for combination := range tasks {
		product := graph.CalculateGroupProduct(combination[:])
		results <- [2]interface{}{combination, product}
		atomic.AddInt32(progress, 1) // Update progress counter
	}
}

func main() {
	// Load graph from file
	filePath := "25-input.txt"
	graph := ReadGraphFromFile(filePath)
	fmt.Println("File loaded and graph constructed successfully!")

	// Get all edges
	var edges [][2]string
	for node, neighbors := range graph.Edges {
		for neighbor := range neighbors {
			if node < neighbor {
				edges = append(edges, [2]string{node, neighbor})
			}
		}
	}

	// Generate combinations
	combinations := GenerateEdgeCombinations(edges)
	totalCombinations := len(combinations)
	fmt.Printf("Total combinations to evaluate: %d\n", totalCombinations)

	// Setup worker pool
	numWorkers := 12 // Set for a 12-core CPU
	tasks := make(chan [3][2]string, totalCombinations)
	results := make(chan [2]interface{}, totalCombinations)

	var wg sync.WaitGroup
	var progress int32

	// Start workers
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			worker(graph, tasks, results, &progress)
		}()
	}

	// Send tasks
	go func() {
		for _, combination := range combinations {
			tasks <- combination
		}
		close(tasks)
	}()

	// Monitor progress
	go func() {
		for {
			processed := atomic.LoadInt32(&progress)
			fmt.Printf("\rProgress: %d/%d combinations processed", processed, totalCombinations)
			if int(processed) >= totalCombinations {
				break
			}
		}
	}()

	// Wait for workers to complete
	go func() {
		wg.Wait()
		close(results)
	}()

	// Find the best result
	maxProduct := 0
	var bestCombination [3][2]string
	for result := range results {
		combination := result[0].([3][2]string)
		product := result[1].(int)
		if product > maxProduct {
			maxProduct = product
			bestCombination = combination
		}
	}

	fmt.Printf("\nBest edges to remove: %v\n", bestCombination)
	fmt.Printf("Maximum product of group sizes: %d\n", maxProduct)
}

File: 23/25/main.py
import networkx as nx
from itertools import combinations
from tqdm import tqdm
from math import comb
from concurrent.futures import ProcessPoolExecutor, as_completed
from multiprocessing import Manager

# Load the wiring diagram and build the graph
def build_graph(file_path):
    graph = nx.Graph()
    with open(file_path, 'r') as f:
        for line in f:
            component, neighbors = line.split(":")
            for neighbor in neighbors.split():
                graph.add_edge(component.strip(), neighbor.strip())
    print("File loaded and graph constructed successfully!")
    return graph

# Calculate product of sizes of two largest connected components
def calculate_group_product(graph, edges_to_remove):
    graph_copy = graph.copy()
    graph_copy.remove_edges_from(edges_to_remove)
    component_sizes = sorted([len(c) for c in nx.connected_components(graph_copy)], reverse=True)
    if len(component_sizes) >= 2:  # Ensure at least two components
        return component_sizes[0] * component_sizes[1]
    return 0

# Function to process a chunk of combinations
def process_chunk(graph_edges, edge_combinations, progress_queue):
    graph = nx.Graph()
    graph.add_edges_from(graph_edges)  # Rebuild graph from edges
    max_product = 0
    best_combination = None
    for edge_combination in edge_combinations:
        product = calculate_group_product(graph, edge_combination)
        if product > max_product:
            max_product = product
            best_combination = edge_combination
        progress_queue.put(1)  # Increment progress
    return best_combination, max_product

# Heuristic: Select top edges by centrality and evaluate with multi-processing
def heuristic_find_best_edges(graph, max_candidates=1000, num_processes=4):
    centrality = nx.edge_betweenness_centrality(graph)
    sorted_edges = sorted(centrality, key=centrality.get, reverse=True)
    candidate_edges = sorted_edges[:max_candidates]
    
    print(f"Testing {len(candidate_edges)} edges using heuristic...")
    total_combinations = comb(len(candidate_edges), 3)
    print(f"Total combinations to evaluate: {total_combinations}")

    # Split combinations into chunks
    edge_combinations = list(combinations(candidate_edges, 3))
    chunk_size = len(edge_combinations) // num_processes
    chunks = [edge_combinations[i:i + chunk_size] for i in range(0, len(edge_combinations), chunk_size)]

    max_product = 0
    best_combination = None

    # Extract edges from the graph for serialization
    graph_edges = list(graph.edges)

    # Progress tracking using a multiprocessing queue
    manager = Manager()
    progress_queue = manager.Queue()

    with tqdm(total=total_combinations, desc="Evaluating combinations") as pbar:
        with ProcessPoolExecutor(max_workers=num_processes) as executor:
            futures = [executor.submit(process_chunk, graph_edges, chunk, progress_queue) for chunk in chunks]
            
            # Monitor progress
            completed = 0
            while completed < total_combinations:
                pbar.update(progress_queue.get())
                completed += 1
            
            for future in as_completed(futures):
                result_combination, result_product = future.result()
                if result_product > max_product:
                    max_product = result_product
                    best_combination = result_combination
    
    return best_combination, max_product

if __name__ == "__main__":
    # Replace with your file path
    file_path = "25-input.txt"
    graph = build_graph(file_path)
    
    # Use heuristic method with multi-processing
    best_edges, max_product = heuristic_find_best_edges(graph, max_candidates=100, num_processes=12)
    
    print("Best edges to remove (heuristic, multi-core):", best_edges)
    print("Maximum product of group sizes (heuristic, multi-core):", max_product)
